src/core/utils/humanize/smartSplit.ts

```diff
@@ -47,70 +69,123 @@ function* smartSplitIterLimit(source: string, limit: number): Generator<string>
-                if (quoteBounds.has(source[i + (inQuote ? 1 : -1)]))
-                    inQuote = !inQuote;
-                else
-                    builder.push(i);
+            case '"':
+                yield* yieldBlock();
+                yield { type: SmartSplitTokenType.QUOTE, start: i, end: i + 1 };
                 break;
-            }
-            case ' ': {
-                if (inQuote)
-                    builder.push(i);
-                else if (start !== undefined) {
-                    yield createSplitItem(source, builder, start, i);
-                    start = undefined;
-                    builder = [];
-                }
+            case '\\':
+                yield* yieldBlock();
+                yield { type: SmartSplitTokenType.ESCAPE, start: i, end: i + 1 };
                 break;
-            }
-            default: {
-                start ??= i;
-                builder.push(i);
+            default:
+                literalStart ??= i;
                 break;
-            }
         }
     }
-    if (start !== undefined)
-        yield createSplitItem(source, builder, start, source.length);
+    yield* yieldBlock();
 }
 
-function createSplitItem(source: string, charIndexes: number[], start: number, end: number): SmartSplitItem {
-    const ranges = charIndexes.reduce<Array<[number, number]>>((ranges, index) => {
-        const prevRange = ranges[ranges.length - 1] ?? [index, index];
-        if (prevRange[1] === index - 1) {
-            prevRange[1] = index;
-        } else {
-            ranges.push([index, index]);
+function* smartSplitIter(source: string): Generator<SmartSplitItem> {
+    const tokens = [...tokenize(source)];
+    if (tokens.length === 0)
+        return;
+
+    const ctx: SmartSplitTokenContext = {
+        ranges: [],
+        index: 0,
+        source,
+        tokenCount: tokens.length,
+        getToken(offset = 0) {
+            const index = this.index + offset;
+            if (index < 0 || index >= tokens.length)
+                return { type: SmartSplitTokenType.BREAK, start: index, end: index };
+            return tokens[index];
+        },
+        * digestRanges() {
+            if (this.ranges.length === 0)
+                return;
+
+            const ranges = this.ranges.splice(0, this.ranges.length);
+            yield {
+                start: Math.min(...ranges.map(t => t.start)),
+                end: Math.min(...ranges.map(t => t.end)),
+                ranges: ranges,
+                content: ranges.map(r => r.content).join('')
+            };
         }
-        return ranges;
-    }, [])
-        .map(([start, end]) => ({
-            start,
-            end,
-            content: source.slice(start, end + 1)
-        }));
-
-    return {
-        ranges: ranges,
-        content: ranges.map(r => r.content).join(''),
-        end,
-        start
     };
+
+    for (; ctx.index < ctx.tokenCount; ctx.index++)
```

Comment:
This loop feels like a hack🤭 Wouldn't be better to just use a while?

Reply:
I did have it as a while at first, but it is more akin to a for loop. you have a counter which increments each loop, and a limit on that counter. The only reason we dont have a `var i = 0` bit is because the loop index can be changed by the method it calls so the value needs to be stored on a reference type